core/shared/src/main/scala/zio/ZRef.scala

```diff
@@ -0,0 +1,238 @@
+/*
+ * Copyright 2017-2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio
+
+import java.util.concurrent.atomic.AtomicReference
+
+/**
+ * A mutable atomic reference for the `IO` monad. This is the `IO` equivalent of
+ * a volatile `var`, augmented with atomic operations, which make it useful as a
+ * reasonably efficient (if low-level) concurrency primitive.
+ *
+ * {{{
+ * for {
+ *   ref <- Ref.make(2)
+ *   v   <- ref.update(_ + 3)
+ *   _   <- console.putStrLn("Value = " + v) // Value = 5
+ * } yield ()
+ * }}}
+ */
+trait ZRef[+EA, +EB, A, +B] extends Serializable { self =>
+
+  /**
+   * Reads the value from the `Ref`.
+   */
+  def get: IO[EB, B]
+
+  /**
+   * Writes a new value to the `Ref`, with a guarantee of immediate
+   * consistency (at some cost to performance).
+   */
+  def set(a: A): IO[EA, Unit]
+
+  /**
+   * Writes a new value to the `Ref` without providing a guarantee of
+   * immediate consistency.
+   */
+  def setAsync(a: A): IO[EA, Unit]
+
+  final def collect[C](pf: PartialFunction[B, C]): ZRef[EA, Option[EB], A, C] =
+    new ZRef[EA, Option[EB], A, C] {
+      def set(a: A): IO[EA, Unit] = self.set(a)
+      def get: IO[Option[EB], C] = self.get.foldM(
+        error => ZIO.fail(Some(error)),
+        value => pf.lift(value).fold[IO[Option[EB], C]](ZIO.fail(None))(ZIO.succeed(_))
+      )
+      def setAsync(a: A): zio.IO[EA, Unit] = self.setAsync(a)
+      private[zio] def unsafeUpdate[E](f: Any => Any): zio.IO[E, C] =
+        self.unsafeUpdate(f).foldM(error => ZIO.fail(error), value => ZIO.succeed(f(value).asInstanceOf[C]))
+      private[zio] def unsafeModify[E, D](f: Any => (Any, Any)): zio.IO[E, D] = self.unsafeModify(f)
+    }
+
+  final def filter(f: B => Boolean): ZRef[EA, Option[EB], A, B] =
+    collect { case b if f(b) => b }
+
+  final def zip[EA1 >: EA, EB1 >: EB, A1, B1](that: ZRef[EA1, EB1, A1, B1]): ZRef[EA1, EB1, (A, A1), (B, B1)] =
```

Comment:
I think `zip` could only be done sanely for `TRef`, not `Ref`, because of the cross-ref transactional guarantees you'd need...

Reply:
👍